### PR TITLE
Add a general search-entries host tool over the entry endpoint

### PR DESCRIPTION
**Pull request by FadhlanR**
## Background and Goal

The AI's search capability is split across two narrow host tools that each drop something discovery flows need: `SearchCardsByQueryTool` searches all available realms with a pinned card scope and returns only id/title summaries, and `SearchCardsByTypeAndTitleTool` cannot express a `specType` or full-text `matches` filter at all. This PR adds one general `search-entries` host tool — a thin interface over the `/_federated-search` entry endpoint (query + realms + scope + limit in, trimmed entry rows out) — and reroutes the two existing tools' descriptions so the model uses them only when it needs live instances to attach, open, or patch.

## Where to start

- `packages/host/app/tools/search-entries.ts` — the tool: scope validation, mixed-scope dedup, relevance-sort composition, wire translation, flattening, and the result mapping.
- `packages/base/commands/search-entry-result.gts` — the input/result card types (re-exported from `command.gts`).
- `packages/host/tests/integration/tools/search-entries-tool-test.gts` — behavior coverage plus unit-style asserts on the exported query-composition helper.

## Key decisions and non-obvious mechanics

- **Transport is `store.searchEntries`, not `store.search`** — the instances path pins `scope: 'cards'` and a data-only fieldset; `searchEntries` passes scope/fields/page through untouched and already handles auth, federation, and 429 shed-retries.
- **Fixed sparse projection** (`item.cardTitle`, `item.cardDescription`, `item.specType`, `item.readMe`, `item.ref`, `item.name`): the fieldset stays internal so a caller can never request full serializations. `name` is the file-meta display handle, so a file row carries more than its bare URL. `readMe` returns in full; the size budget is the `limit` input (default 5, max 10). Tool result cards are inlined verbatim into every subsequent prompt turn with no downstream truncation, so the tool bounds its own output — and the result card uses `containsMany` summaries, never `linksToMany(CardDef)`.
- **Default scope `'all'` with automatic dedup**: a card matches both its instance row and its dual-indexed `.json` file row under the mixed scope, so the filter gains `excludeCardInstanceFileRows()` unless it already carries a kind-narrowing positive type ref (`hasNarrowingPositiveTypeRef`, now exported from the host search query-builder rather than copied a third time). Narrowing is structure-aware: one narrowing branch narrows an `every`, but an `any` narrows only when every branch does — a disjunction with an unanchored branch can still match both row kinds.
- **Relevance**: the `_matchRelevance` sort is appended only when the filter has a positive `matches` term and no explicit sort (the server 400s the sort without such a term). Merged rows are re-sorted client-side because the federated merge concatenates per realm without cross-realm re-ranking — but only when the tool added the sort itself (`composeSearchEntriesQuery` reports the addition), so a caller's own ordering, which may legally include `_matchRelevance`, is never overridden.
- **Partial results are labelled, not hidden**: a realm that fails during the federated fan-out doesn't throw — the merge marks the document `meta.incomplete` and returns the realms that answered. The result card carries that as an `incomplete` field and the tool description tells the model it means matches may be missing and `total` undercounts.
- **The tool module default-exports the class** so skill frontmatter can declare it as `module: '@cardstack/boxel-host/tools/search-entries', name: default`. The skill-side declaration lands separately in boxel-skills.
- Errors propagate to the model instead of collapsing to an empty result, so it can correct query grammar.

## Testing

- New integration suite: 17 tests covering the projection (including the file-row `name`), scope narrowing and validation, mixed-scope dedup (the narrowing-anchor skip and the partially-anchored `any` non-skip), relevance sorting, its absence without `matches`, the caller-sort passthrough, limit defaulting/clamping, realm targeting, the `incomplete` flag default, and the default-export guard.
- Existing `tools | search` suite, the `search/query-builder` unit suite, and the host tool schema-generation suite (which now includes the new tool via `HostToolClasses`) pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/host/app/tools/search-entries.ts [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+      rows = [...rows].sort(
+        (a, b) => (b.matchRelevance ?? -1) - (a.matchRelevance ?? -1),
+      );
+    }
+
+    let commandModule = await this.loadToolModule();
+    let { SearchEntriesResult, SearchEntrySummaryField } = commandModule;
+    return new SearchEntriesResult({
+      results: rows.map((row) => new SearchEntrySummaryField(row)),
+      total: doc.meta.page.total,
```
Comment (FadhlanR) [inline packages/host/app/tools/search-entries.ts]: [Claude Code 🤖] A partial federated result reads as complete here. When one of the fanned-out realms fails or never resolves to an instance, the server's merge stamps `meta.incomplete: true` on the document (`combineSearchEntryResults` in `runtime-common/search-entry.ts`) and returns the realms that answered — the tool drops that flag, so the model sees a shorter list and an undercounted `total` with nothing marking them as partial. This is the one error path that does not propagate as a thrown error, so the errors-propagate-to-the-model principle doesn't cover it. Suggest carrying it through: an `incomplete` field on `SearchEntriesResult` plus a sentence in the tool description telling the model what it means.

Gap in this PR's new surface; non-blocking.

File: packages/host/app/tools/search-entries.ts [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    let doc = await this.store.searchEntries(wireQuery, realms);
+
+    let rows = summarizeEntries(doc);
+    if (rows.some((row) => row.matchRelevance !== undefined)) {
+      // The federated merge concatenates per-realm results without re-ranking
+      // across realms; relevance rides each entry so the merged page can be.
+      rows = [...rows].sort(
+        (a, b) => (b.matchRelevance ?? -1) - (a.matchRelevance ?? -1),
+      );
+    }
```
Comment (FadhlanR) [inline packages/host/app/tools/search-entries.ts]: [Claude Code 🤖] This re-sort can clobber a caller's explicit sort. It keys on relevance riding the rows, but a caller may legally put `_matchRelevance` in its own `sort` (the wire grammar accepts it whenever the filter has a positive `matches` term) — e.g. `[{by: 'cardTitle'}, {by: '_matchRelevance'}]`, or an `asc` direction — and the server then stamps `meta._matchRelevance` on every row, so this block replaces that ordering with relevance-desc. Keying the re-sort on "`composeSearchEntriesQuery` added the sort" (return that fact from the helper) fixes the clobber and gives the cross-realm merge re-sort a unit-testable seam. Worth having: the integration relevance test cannot catch this block regressing, because a single-realm fixture arrives already server-sorted, so the sorted-order assert passes with or without the re-sort.

Edge case in new code; non-blocking.

File: packages/host/app/tools/search-entries.ts [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+// judge a candidate rides the row itself (including the full readMe), so no
+// per-hit follow-up read is required; the fieldset stays internal so a caller
+// can never request full serializations that blow up the result size.
+const PROJECTION_FIELDS = [
+  'item.cardTitle',
+  'item.cardDescription',
+  'item.specType',
+  'item.readMe',
+  'item.ref',
+];
```
Comment (FadhlanR) [inline packages/host/app/tools/search-entries.ts]: [Claude Code 🤖] Under this projection a file row surfaces with every summary field empty: the sparse fieldset intersects with the item's attributes, and a `file-meta` resource carries `name`/`contentType`/… but none of these five (`fileResourceFromIndex` + `buildSparseItemResource`). A `scope: 'files'` discovery result is therefore a bare URL list, and the embedded result template renders just `(url)` per row. Adding `item.name` here (and arguably `item.contentType`) plus a matching summary field would give file rows a display handle; on card rows the extra field is a no-op since the attribute is absent.

Non-blocking.

File: packages/host/app/tools/search-entries.ts [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+// unless it already carries a kind-narrowing positive type ref, which matches
+// only one of the two. A filter with a positive full-text `matches` term and
+// no explicit sort gains the relevance sort (the server rejects that sort
+// without such a term, so it is never added unconditionally).
+export function composeSearchEntriesQuery(
+  query: Query,
+  scope: SearchEntryScope,
+): Query {
+  let filter = query.filter;
+  if (scope === 'all' && !hasNarrowingPositiveTypeRef(filter)) {
```
Comment (FadhlanR) [inline packages/host/app/tools/search-entries.ts]: [Claude Code 🤖] Pre-existing, now load-bearing: `hasNarrowingPositiveTypeRef` counts one narrowing branch of an `any` as narrowing the whole filter (`getTypeRefsFromFilter` collects across `any` children), so a filter like `{any: [{type: <card type>}, {matches: '…'}]}` at the default scope skips the dedup wrap, and the `matches` branch can return a card twice — its instance row plus its `.json` file row, burning limit slots. The search sheet only ever feeds this helper controlled shapes; this tool is the first surface handing it arbitrary model-authored filters. Not this PR's bug — a follow-up on the helper (treat `any` as narrowing only when every branch narrows) would cover both consumers.

Follow-up; non-blocking.

**Review by FadhlanR — COMMENTED** [review]
[Claude Code 🤖] (withdrawn)

**Comment by FadhlanR** [comment]
[Claude Code 🤖] Reviewer heads-up: in the Files changed view the diff may look like it has an indentation problem across all the files. It's a GitHub diff-rendering artifact, not a real issue and not specific to any file type — the files are all consistent 2-space indentation with no tabs and no unusual whitespace. Safe to ignore.

File: packages/host/app/tools/search-entries.ts [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+    let { query, addedRelevanceSort } = composeSearchEntriesQuery(
+      input.query,
+      scope,
+    );
+    let wireQuery = searchEntryWireQueryFromQuery(query, {
+      fields: PROJECTION_FIELDS,
+      scope,
+    });
+    wireQuery.page = { ...wireQuery.page, size: limit };
```
Comment (richardhjtan) [inline packages/host/app/tools/search-entries.ts:198]: [Claude Code 🤖]

This page rides *each* realm's own search. `combineSearchEntryResults` in `search-entry.ts` concatenates every realm's `data` with no truncation, and `run()` never slices `rows` — so with `realms` omitted (the documented default: every realm the caller can read) a `limit: 10` returns up to 10 × realmCount summaries, each carrying a full `readMe`. That result size is what the fixed `PROJECTION_FIELDS` comment names as the reason the fieldset stays internal.

The relevance re-sort below inherits the same gap: it re-ranks the merged rows without truncating, so it reorders rows the caller was never meant to receive rather than picking the top N.

`rows = rows.slice(0, limit)` closes the case the tool sorts itself. The caller-supplied-sort case needs a decision first — slicing a realm-concatenated list drops whole realms arbitrarily — so either re-sort unconditionally before slicing, or document `limit` as per-realm and stop promising a maximum in the description.

Regression, blocking.

File: packages/host/tests/integration/tools/search-entries-tool-test.gts [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+    });
+    assert.ok(
+      withoutMatches.results.every(
+        (r: { matchRelevance?: number }) => r.matchRelevance == null,
+      ),
+      'no relevance without a matches term (and no 400 from the sort)',
+    );
+  });
+
+  test('limit defaults to 5, is honored, and clamps at 10', async function (assert) {
```
Comment (richardhjtan) [inline packages/host/tests/integration/tools/search-entries-tool-test.gts:309]: [Claude Code 🤖]

This runs against a single realm, where the server's per-realm page bound and the tool's cap are the same number — so it passes whether or not the cap holds, and cannot fail on the case that breaks it. Two realms with five matches each, asserting `results.length === 5` at `limit: 5`, is what would pin it.

Test coverage, blocking alongside the fix.

**Review by richardhjtan — COMMENTED** [review]
[Claude Code 🤖]

Reviewed the tool against the entry API's federated fan-out.

**Blocking: `limit` caps each realm, not the result.** Threads on `search-entries.ts` and the test that reads as pinning the cap.

File: packages/host/app/tools/search-entries.ts (hunk 5 above)
Comment (FadhlanR) [reply packages/host/app/tools/search-entries.ts:198]: Confirmed and fixed in 7beefaf — thanks, this was real.

Verified the mechanism: `combineSearchEntryResults` does `combined.data.push(...doc.data)` with no truncation, and `run()` never sliced, so `page.size` was bounding each realm rather than the result while the description promised a maximum.

It also reproduced outside the tests. In an assistant session, a search with no `realms` and no `limit` — so the default of 5 — returned **10 rows** across three realms (5 from the catalog, 4 from one user realm, 1 from another). Exactly the shape you described.

**On the two options you offered, I went with neither as stated.**

Documenting `limit` as per-realm doesn't work for the caller: the model can't compute a ceiling that depends on how many realms it can read, and that count varies per user and workspace. It would also undercut the reason the projection is fixed — bounding response size — since the bound would multiply by an unknown factor.

Re-sorting unconditionally isn't implementable either. We project six fields, and a caller may legitimately sort on a key we don't return, so there's nothing to re-rank on client-side.

So: **slice always**, placed after the tool's own relevance re-sort so that path still picks the true top N. `total` keeps reporting the full summed count, which is what signals the page is partial.

What I couldn't preserve is global ordering for a caller-supplied sort across realms — that's inherent to the fan-out, and the old code only hid it by returning everything. Rather than over-promise, the description now says what holds:

> Relevance-ranked results are ordered across all realms; a sort you supply yourself orders rows within each realm and then merges them.

If we want globally-ordered caller sorts, that belongs in the entry endpoint rather than the tool, and is bigger than this PR.

File: packages/host/tests/integration/tools/search-entries-tool-test.gts (hunk 6 above)
Comment (FadhlanR) [reply packages/host/tests/integration/tools/search-entries-tool-test.gts:309]: Right on both counts — added in 7beefaf, in the shape you specified.

`limit caps the merged result, not each realm`: a second realm at `test2/` with five matching authors alongside five in the first realm, asserting `results.length === 5` at `limit: 5` across both, plus `total > 5` so truncation stays visible.

**Verified it fails without the fix**, which was your actual point about the original:

| | with `slice` | without `slice` |
|---|---|---|
| `limit caps the merged result` | pass | **fail — expected 5, got 10** |

10 being 2 realms × the per-realm page of 5.

One wrinkle worth recording. My first version of the fixture gave the second realm's authors `lastName: 'Example'` — the same surname the existing default-limit test counts. That test searches all realms, so its `total` went 12 → 17 and it went red. Both realms now use `lastName: 'Capped'`, queried by the federated test alone, leaving the 'Example' population at 12.

Full `Integration | tools | search-entries` module: 18 tests, 44/44 assertions, 0 failed.
